# Análise de Risco de Crédito - NuBank

Risco de crédito está associado à possibilidade de um cliente não cumprir com as obrigações contratuais, como hipotecas, dívidas de cartão de crédito e outros tipos de empréstimos.

Minimizar o risco de inadimplência é uma grande preocupação para instituições financeiras. Por esse motivo, bancos comerciais e de investimento, fundos de capital de risco, empresas de gestão de ativos e seguradoras, para citar alguns, estão cada vez mais contando com a tecnologia para prever quais clientes são mais propensos a não honrar com as suas dívidas.

Modelos de Machine Learning têm ajudado essas empresas a melhorar a precisão de suas análises de risco de crédito, fornecendo um método científico para identificar devedores em potencial com antecedência.

Neste projeto, construiremos um modelo para prever o risco de inadimplência do cliente para o Nubank, uma das maiores e importantes Fintechs brasileira.

## Importação das Bibliotecas

In [158]:
# import dos pacotes
import pandas as pd
import numpy as np
import matplotlib as mpl
import matplotlib.pyplot as plt
import seaborn as sns

# definição do estilo estético das plotagens
sns.set_style()

# filtragem de warnings
import warnings
warnings.filterwarnings('ignore')   

from sklearn.impute import SimpleImputer

from sklearn.model_selection import train_test_split

# Set the option to display all columns
pd.set_option('display.max_columns', None)

# Seaborn grid style.
sns.set_theme(style='whitegrid')



## Importação do Dataset

In [159]:
df = pd.read_csv('data/acquisition_train.csv')

## Visualização e Entendimento dos Dados

In [160]:
df.head()

ids target_default  \
0  343b7e7b-2cf8-e508-b8fd-0a0285af30aa          False   
1  bc2c7502-bbad-0f8c-39c3-94e881967124          False   
2  669630dd-2e6a-0396-84bf-455e5009c922           True   
3  d235609e-b6cb-0ccc-a329-d4f12e7ebdc1          False   
4  9e0eb880-e8f4-3faa-67d8-f5cdd2b3932b          False   

                    score_1                   score_2  score_3     score_4  \
0  1Rk8w4Ucd5yR3KcqZzLdow==  IOVu8au3ISbo6+zmfnYwMg==    350.0  101.800832   
1  DGCQep2AE5QRkNCshIAlFQ==  SaamrHMo23l/3TwXOWgVzw==    370.0   97.062615   
2  DGCQep2AE5QRkNCshIAlFQ==  Fv28Bz0YRTVAT5kl1bAV6g==    360.0  100.027073   
3  1Rk8w4Ucd5yR3KcqZzLdow==  dCm9hFKfdRm7ej3jW+gyxw==    510.0  101.599485   
4  8k8UDR4Yx0qasAjkGrUZLw==  +CxEO4w7jv3QPI/BQbyqAA==    500.0   98.474289   

    score_5     score_6  risk_rate  last_amount_borrowed  \
0  0.259555  108.427273       0.40              25033.92   
1  0.942655   92.002546       0.24                   NaN   
2  0.351918  112.892453       0.29               7207.92   
3  0.987673   94.902491       0.32                   NaN   
4  0.532539  118.126207       0.18                   NaN   

   last_borrowed_in_months  credit_limit  \
0                     36.0           0.0   
1                      NaN       39726.0   
2                     36.0           NaN   
3                      NaN       54591.0   
4                      NaN           NaN   

                                              reason     income  \
0  mLVIVxoGY7TUDJ1FyFoSIZi1SFcaBmO01AydRchaEiGYtU...   65014.12   
1  mLVIVxoGY7TUDJ1FyFoSIZi1SFcaBmO01AydRchaEiGYtU...  100018.91   
2  mLVIVxoGY7TUDJ1FyFoSIZi1SFcaBmO01AydRchaEiGYtU...   65023.65   
3  mLVIVxoGY7TUDJ1FyFoSIZi1SFcaBmO01AydRchaEiGYtU...   68830.01   
4  mLVIVxoGY7TUDJ1FyFoSIZi1SFcaBmO01AydRchaEiGYtU...   60011.29   

  facebook_profile                     state                       zip  \
0             True  sjJbkqJS7cXalHLBFA+EOQ==  Ernn+uVXCMq/6ARrBCcd+A==   
1            False  xsd3ZdsI3356I3xMxZeiqQ==  rlWIXTBO+VOa34+SpGyhlQ==   
2              NaN  Ygq6MsM98oC8yceExr69Ig==  PjTIDfJsK0DKL9fO7vuW2g==   
3            False  KET/Pmr6rHp1RJ/P9ymztw==  Cc/kWDLQH3dpHv5HU+pLVA==   
4             True  xsd3ZdsI3356I3xMxZeiqQ==  i036nmJ7rfxo+3EvCD7Jnw==   

                    channel  \
0  NCqL3QBx0pscDnx3ixKwXg==   
1  NCqL3QBx0pscDnx3ixKwXg==   
2  NCqL3QBx0pscDnx3ixKwXg==   
3  NCqL3QBx0pscDnx3ixKwXg==   
4  NCqL3QBx0pscDnx3ixKwXg==   

                                            job_name  \
0  mLVIVxoGY7TUDJ1FyFoSIZi1SFcaBmO01AydRchaEiGYtU...   
1  mLVIVxoGY7TUDJ1FyFoSIZi1SFcaBmO01AydRchaEiGYtU...   
2  mLVIVxoGY7TUDJ1FyFoSIZi1SFcaBmO01AydRchaEiGYtU...   
3  mLVIVxoGY7TUDJ1FyFoSIZi1SFcaBmO01AydRchaEiE56f...   
4  mLVIVxoGY7TUDJ1FyFoSIZi1SFcaBmO01AydRchaEiGYtU...   

                 real_state  ok_since  n_bankruptcies  n_defaulted_loans  \
0  N5/CE7lSkAfB04hVFFwllw==       NaN             0.0                0.0   
1  N5/CE7lSkAfB04hVFFwllw==      50.0             0.0                0.0   
2  n+xK9CfX0bCn77lClTWviw==       NaN             0.0                0.0   
3  n+xK9CfX0bCn77lClTWviw==       NaN             1.0                0.0   
4  n+xK9CfX0bCn77lClTWviw==       NaN             0.0                0.0   

   n_accounts  n_issues application_time_applied  application_time_in_funnel  \
0        18.0      18.0                 07:52:34                         444   
1        14.0      14.0                 02:34:29                         346   
2        10.0       NaN                 00:60:02                           6   
3        19.0      19.0                 11:20:49                         406   
4        11.0       NaN                 13:39:03                         240   

         email  external_data_provider_credit_checks_last_2_year  \
0  outlook.com                                               NaN   
1    gmail.com                                               0.0   
2    gmail.com                                               NaN   
3    spgov

In [161]:
# Dataframe shape
print('Número de linhas do dataset: ', df.shape[0])
print('Número de colunas do dataset: ', df.shape[1])

Número de linhas do dataset:  45000
Número de colunas do dataset:  43


In [162]:
# Informações sobre o dataset	
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 45000 entries, 0 to 44999
Data columns (total 43 columns):
 #   Column                                            Non-Null Count  Dtype  
---  ------                                            --------------  -----  
 0   ids                                               45000 non-null  object 
 1   target_default                                    41741 non-null  object 
 2   score_1                                           44438 non-null  object 
 3   score_2                                           44438 non-null  object 
 4   score_3                                           44438 non-null  float64
 5   score_4                                           45000 non-null  float64
 6   score_5                                           45000 non-null  float64
 7   score_6                                           45000 non-null  float64
 8   risk_rate                                         44438 non-null  float64
 9   last_amount_borro

- Podemos ver que algumas variáveis possuem valores ausentes. Vamos dar uma olhada mais de perto nelas.

In [163]:
# Porcentagem de valores faltantes
print(((df.isnull().sum() / df.shape[0]) * 100).sort_values(ascending=False))

target_fraud                                        96.617778
last_amount_borrowed                                66.568889
last_borrowed_in_months                             66.568889
ok_since                                            58.988889
external_data_provider_credit_checks_last_2_year    50.284444
external_data_provider_credit_checks_last_year      33.608889
credit_limit                                        30.666667
n_issues                                            25.653333
facebook_profile                                     9.906667
marketing_channel                                    7.951111
job_name                                             7.413333
target_default                                       7.242222
external_data_provider_email_seen_before             4.962222
lat_lon                                              3.028889
user_agent                                           1.604444
n_bankruptcies                                       1.548889
n_defaul

- Verificar as estatísticas descritivas das variáveis numéricas.

In [164]:
df.describe().T

count          mean  \
score_3                                           44438.0  3.479833e+02   
score_4                                           45000.0  1.000068e+02   
score_5                                           45000.0  5.000371e-01   
score_6                                           45000.0  9.992523e+01   
risk_rate                                         44438.0  2.967172e-01   
last_amount_borrowed                              15044.0  1.378234e+04   
last_borrowed_in_months                           15044.0  4.176389e+01   
credit_limit                                      31200.0  3.438195e+04   
income                                            44438.0  7.160739e+04   
ok_since                                          18455.0  3.506974e+01   
n_bankruptcies                                    44303.0  7.627023e-02   
n_defaulted_loans                                 44426.0  4.569396e-03   
n_accounts                                        44438.0  1.069670e+01   
n_issues                                          33456.0  1.106812e+01   
application_time_in_funnel                        45000.0  2.475370e+02   
external_data_provider_credit_checks_last_2_year  22372.0  0.000000e+00   
external_data_provider_credit_checks_last_month   45000.0  1.501067e+00   
external_data_provider_credit_checks_last_year    29876.0  5.033806e-01   
external_data_provider_email_seen_before          42767.0  1.253478e+01   
external_data_provider_fraud_score                45000.0  5.003998e+02   
reported_income                                   45000.0           inf   
shipping_zip_code                                 45000.0  3.392351e+04   

                                                           std           min  \
score_3                                             110.429165      0.000000   
score_4                                               3.178080     86.191572   
score_5                                               0.288326      0.000035   
score_6                                              10.016549     60.663039   
risk_rate                                             0.101933      0.000000   
last_amount_borrowed                               8077.571933   1005.180000   
last_borrowed_in_months                              10.252705     36.000000   
credit_limit                                      36586.982218      0.000000   
income                                            51858.967748   4821.180000   
ok_since                                             21.663194      0.000000   
n_bankruptcies                                        0.274052      0.000000   
n_defaulted_loans                                     0.079400      0.000000   
n_accounts                                            4.592463      0.000000   
n_issues                                              4.595501      0.000000   
application_time_in_funnel                          146.118398      0.000000   
external_data_provider_credit_checks_last_2_year      0.000000      0.000000   
external_data_provider_credit_checks_last_month       1.114941      0.000000   
external_data_provider_credit_checks_last_year        0.499997      0.000000   
external_data_provider_email_seen_before            126.530487   -999.000000   
external_data_provider_fraud_score                  287.968873      0.000000   
reported_income                                            NaN    403.000000   
shipping_zip_code                                 12692.406382  12001.000000   

                                                           25%            50%  \
score_3                                             270.000000     340.000000   
score_4                                              97.867036     100.018900   
score_5                                               0.252150       0.500753   
score_6                                              93.198353      99.974796   
risk_rate                                             0.220000       0.290000   

## Relatório Análise Inicial dos Dados

- O dataset possui **43** variáveis e **45.000** registros.
- Dessas **43** variáveis, **8** possuem mais de **10%** de valores missing, **target_fraud** é a variável com a maior porcentagem de valores ausentes **(96.6%)**, em segundo e terceiro lugar temos as variáveis **last_amount_borrowed** e **last_borrowed_in_months** com **(66.5%)** de valores ausentes.
- A variável **target_fraud** mostra se existe ou não fraude de cartão de crédito. Nesse caso essa variável não é importante para o projeto.Portanto, esta coluna será excluída.
- As variáveis **last_amount_borrowed**, **last_borrowed_in_months**, **ok_since** e **external_data_provider_credit_checks_last_2_year** possuem mais de 50% de valores ausentes, substituir esses valores é algo difícil, manter elas pode gerar erros durante a predição. Portanto, estas colunas serão excluídas.
- As variáveis **job_name**, **external_data_provider_first_name**, **profile_phone_number** e **zip** são variáveis insignificantes para o projeto, portanto serão excluídas.
- A variável **external_data_provider_email_seen_before** possui um valor mínimo de -999.000000, certamente foi um erro, nesse caso, vamos alterar para o tipo NaN.
- A variável **reported_income** possui um valor máximo descrito como inf(infinito), nesse caso, vamos alterar para o tipo NaN.
- A variável **credit_limit** possui um valor mínimo de 0.000000, isso não existe em instituições financeiras, é obrigatório liberar um valor X de crédito para o cliente. Portanto, este valor será substituído por NaN.
- As variáveis:
    - **ids**, **score_1**, **score_2**, **reason**, **facebook_profile**, **state**, **zip**, **channel**, **job_name**, **real_state**,
**email**, **external_data_provider_first_name**, **lat_lon**, **marketing_channel**, 
**profile_phone_number**, **shipping_zip_code**, **profile_tags**, **user_agent**, e **target_fraud** são variáveis com valores não significantes para o projeto, portanto serão excluídas.

## Exclusão das Variáveis não significativas

In [165]:
# cópia do dataframe
df2 = df.copy()

In [166]:
exclude_columns = ["ids", "score_1", "score_2", "reason", "facebook_profile", "state", "zip", "channel", "job_name", "real_state",
                    "email", "external_data_provider_first_name", "lat_lon", "marketing_channel", 
                    "profile_phone_number", "shipping_zip_code", "profile_tags", "user_agent", "target_fraud"]

df2.drop(labels = exclude_columns, axis=1, inplace=True)

## Alteração da Variável credit_limit com valor 0

In [167]:
df2['credit_limit'] = df2['credit_limit'].apply(lambda x: np.nan if x == 0 else x)

## Alteração da Variável reported_income com Valor Inf

In [168]:
df2.replace([np.inf, -np.inf], np.nan, inplace=True)

## Divisão dos dados em treino e teste

- Antes de iniciar as transformações e pré-processamentos dos dados, é necessário dividir o dataset em treino e teste, mas porque?
    - Porque precisamos evitar o "Data Leakage". Data Leakage ocorre quando informações do conjunto de dados de teste ou validação vazam para o conjunto de treinamento durante o pré-processamento ou modelagem, ou quando informaçõpes do target vazam para as features. Nessas situações, o que vai acontecer é que você vai ver um modelo muito bom, mas isso será ilusório, pois o seu modelo "roubou" para ter o resultado bom.

- Referências sobre Data Leakage: 
    - https://www.linkedin.com/company/universidade-dos-dados/posts/?feedView=all
    - https://www.casadocodigo.com.br/products/livro-escd
    - https://estatsite.com.br/2020/12/12/data-leakage-o-erro-que-ate-os-grandes-cometem/


In [169]:
# Dividindo os dados primeiro

X = df2.drop(columns=['target_default'])	
y = df2['target_default']

X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.25, random_state=42)

In [170]:
train = pd.concat([X_train, y_train], axis=1)
train.reset_index(drop=True, inplace=True)
train.head()

score_3     score_4   score_5     score_6  risk_rate  last_amount_borrowed  \
0    240.0  105.331178  0.893486   94.495216       0.26                   NaN   
1    270.0  103.211042  0.605574  101.567582       0.33              19237.32   
2    260.0  102.557916  0.917576  117.057610       0.28                   NaN   
3    300.0  101.664122  0.627588   89.140113       0.28                   NaN   
4    350.0  109.603725  0.657344   98.574923       0.27                   NaN   

   last_borrowed_in_months  credit_limit    income  ok_since  n_bankruptcies  \
0                      NaN       21968.0  45013.96       NaN             1.0   
1                     36.0       40972.0  80022.23       NaN             0.0   
2                      NaN           NaN  19225.52       NaN             0.0   
3                      NaN           NaN  60043.78      62.0             NaN   
4                      NaN      156549.0  45032.90      45.0             0.0   

   n_defaulted_loans  n_accounts  n_issues application_time_applied  \
0                0.0        11.0      11.0                 19:41:46   
1                0.0        12.0      12.0                 05:04:32   
2                0.0         7.0       NaN                 20:20:48   
3                0.0         9.0       NaN                 03:43:14   
4                0.0        25.0      25.0                 09:09:36   

   application_time_in_funnel  \
0                         278   
1                         141   
2                         114   
3                         484   
4                         300   

   external_data_provider_credit_checks_last_2_year  \
0                                               0.0   
1                                               0.0   
2                                               NaN   
3                                               0.0   
4                                               NaN   

   external_data_provider_credit_checks_last_month  \
0                                                2   
1                                                3   
2                                                1   
3                                                2   
4                                                3   

   external_data_provider_credit_checks_last_year  \
0                                             NaN   
1                                             1.0   
2                                             NaN   
3                                             NaN   
4                                             NaN   

   external_data_provider_email_seen_before  \
0                                      12.0   
1                                      57.0   
2                                      54.0   
3                                      20.0   
4                                      42.0   

   external_data_provider_fraud_score  reported_income shipping_state  \
0                                 741          92586.0          BR-ES   
1                                 659          95975.0          BR-GO   
2                                 905          53981.0          BR-PR   
3                                 603         140976.0          BR-DF   
4                                 911         120129.0          BR-SP   

  target_default  
0          False  
1          False  
2          False  
3          False  
4          False

## Input Missing

In [171]:
train_num = train.select_dtypes(exclude='object').columns
train_cat = train.select_dtypes(include='object').columns

In [172]:
# Input Missing Variáveis Numéricas

imputer = SimpleImputer(missing_values=np.nan, strategy='median')
imputer = imputer.fit(train.loc[:, train_num])
train.loc[:, train_num] = imputer.transform(train.loc[:, train_num])

# Input Missing Variáveis Categóricas

imputer = SimpleImputer(missing_values=np.nan, strategy='most_frequent')
imputer = imputer.fit(train.loc[:, train_cat])
train.loc[:, train_cat] = imputer.transform(train.loc[:, train_cat])


In [173]:
train.isna().sum()

score_3                                             0
score_4                                             0
score_5                                             0
score_6                                             0
risk_rate                                           0
last_amount_borrowed                                0
last_borrowed_in_months                             0
credit_limit                                        0
income                                              0
ok_since                                            0
n_bankruptcies                                      0
n_defaulted_loans                                   0
n_accounts                                          0
n_issues                                            0
application_time_applied                            0
application_time_in_funnel                          0
external_data_provider_credit_checks_last_2_year    0
external_data_provider_credit_checks_last_month     0
external_data_provider_credi

## Análise Exploratória de Dados

### Entendimento das Variáveis

**Score**

O score de crédito é geralmente calculado com base em modelos estatísticos que analisam o histórico financeiro de uma pessoa ou empresa. Esses modelos consideram uma variedade de informações, incluindo:

- Histórico de pagamento: Frequência com que você paga suas contas no prazo.
- Dívidas em aberto: Valor e número de dívidas existentes.
- Uso do crédito: Proporção de crédito usado em relação ao limite disponível (chamado de "taxa de utilização").
- Tempo de crédito: Quantidade de tempo que você possui histórico financeiro.
- Consultas de crédito: Quantas vezes seu relatório de crédito foi consultado recentemente.
- Registros negativos: Presença de atrasos, inadimplências, protestos ou ações judiciais.

O score de crédito é tipicamente representado por um número dentro de um intervalo, como:
- 300 a 850 (nos EUA, em modelos como FICO ou VantageScore)
- 0 a 1.000 (no Brasil, por birôs como Serasa, Boa Vista ou SPC)

O score brasileiro é geralmente dividido em faixas:

- Baixo (0-300): Alto risco de inadimplência.
- Médio (301-700): Risco moderado.
- Alto (701-1000): Baixo risco de inadimplência.

**Target Default**

Target default no contexto de credit risk (risco de crédito) refere-se à definição de inadimplência que é usada como a variável alvo (target) em modelos de previsão de risco de crédito. Essa variável indica se um cliente (pessoa física ou jurídica) cumpriu ou não suas obrigações financeiras, como pagamentos de empréstimos ou financiamentos.

In [174]:
train.head()

score_3     score_4   score_5     score_6  risk_rate  last_amount_borrowed  \
0    240.0  105.331178  0.893486   94.495216       0.26              12024.02   
1    270.0  103.211042  0.605574  101.567582       0.33              19237.32   
2    260.0  102.557916  0.917576  117.057610       0.28              12024.02   
3    300.0  101.664122  0.627588   89.140113       0.28              12024.02   
4    350.0  109.603725  0.657344   98.574923       0.27              12024.02   

   last_borrowed_in_months  credit_limit    income  ok_since  n_bankruptcies  \
0                     36.0       21968.0  45013.96      32.0             1.0   
1                     36.0       40972.0  80022.23      32.0             0.0   
2                     36.0       31510.0  19225.52      32.0             0.0   
3                     36.0       31510.0  60043.78      62.0             0.0   
4                     36.0      156549.0  45032.90      45.0             0.0   

   n_defaulted_loans  n_accounts  n_issues application_time_applied  \
0                0.0        11.0      11.0                 19:41:46   
1                0.0        12.0      12.0                 05:04:32   
2                0.0         7.0      10.0                 20:20:48   
3                0.0         9.0      10.0                 03:43:14   
4                0.0        25.0      25.0                 09:09:36   

   application_time_in_funnel  \
0                         278   
1                         141   
2                         114   
3                         484   
4                         300   

   external_data_provider_credit_checks_last_2_year  \
0                                               0.0   
1                                               0.0   
2                                               0.0   
3                                               0.0   
4                                               0.0   

   external_data_provider_credit_checks_last_month  \
0                                                2   
1                                                3   
2                                                1   
3                                                2   
4                                                3   

   external_data_provider_credit_checks_last_year  \
0                                             1.0   
1                                             1.0   
2                                             1.0   
3                                             1.0   
4                                             1.0   

   external_data_provider_email_seen_before  \
0                                      12.0   
1                                      57.0   
2                                      54.0   
3                                      20.0   
4                                      42.0   

   external_data_provider_fraud_score  reported_income shipping_state  \
0                                 741          92586.0          BR-ES   
1                                 659          95975.0          BR-GO   
2                                 905          53981.0          BR-PR   
3                                 603         140976.0          BR-DF   
4                                 911         120129.0          BR-SP   

  target_default  
0          False  
1          False  
2          False  
3          False  
4          False

In [175]:
train.to_csv('data/train.csv', index=False)

In [176]:
train.head()

score_3     score_4   score_5     score_6  risk_rate  last_amount_borrowed  \
0    240.0  105.331178  0.893486   94.495216       0.26              12024.02   
1    270.0  103.211042  0.605574  101.567582       0.33              19237.32   
2    260.0  102.557916  0.917576  117.057610       0.28              12024.02   
3    300.0  101.664122  0.627588   89.140113       0.28              12024.02   
4    350.0  109.603725  0.657344   98.574923       0.27              12024.02   

   last_borrowed_in_months  credit_limit    income  ok_since  n_bankruptcies  \
0                     36.0       21968.0  45013.96      32.0             1.0   
1                     36.0       40972.0  80022.23      32.0             0.0   
2                     36.0       31510.0  19225.52      32.0             0.0   
3                     36.0       31510.0  60043.78      62.0             0.0   
4                     36.0      156549.0  45032.90      45.0             0.0   

   n_defaulted_loans  n_accounts  n_issues application_time_applied  \
0                0.0        11.0      11.0                 19:41:46   
1                0.0        12.0      12.0                 05:04:32   
2                0.0         7.0      10.0                 20:20:48   
3                0.0         9.0      10.0                 03:43:14   
4                0.0        25.0      25.0                 09:09:36   

   application_time_in_funnel  \
0                         278   
1                         141   
2                         114   
3                         484   
4                         300   

   external_data_provider_credit_checks_last_2_year  \
0                                               0.0   
1                                               0.0   
2                                               0.0   
3                                               0.0   
4                                               0.0   

   external_data_provider_credit_checks_last_month  \
0                                                2   
1                                                3   
2                                                1   
3                                                2   
4                                                3   

   external_data_provider_credit_checks_last_year  \
0                                             1.0   
1                                             1.0   
2                                             1.0   
3                                             1.0   
4                                             1.0   

   external_data_provider_email_seen_before  \
0                                      12.0   
1                                      57.0   
2                                      54.0   
3                                      20.0   
4                                      42.0   

   external_data_provider_fraud_score  reported_income shipping_state  \
0                                 741          92586.0          BR-ES   
1                                 659          95975.0          BR-GO   
2                                 905          53981.0          BR-PR   
3                                 603         140976.0          BR-DF   
4                                 911         120129.0          BR-SP   

  target_default  
0          False  
1          False  
2          False  
3          False  
4          False